internal/functions/efgs/upload-batch.go

```diff
@@ -0,0 +1,159 @@
+package efgs
+
+import (
+	"bytes"
+	"context"
+	b64 "encoding/base64"
+	"encoding/json"
+	"errors"
+	efgsutils "github.com/covid19cz/erouska-backend/internal/functions/efgs/utils"
+	"github.com/covid19cz/erouska-backend/internal/logging"
+	"github.com/covid19cz/erouska-backend/internal/secrets"
+	"github.com/golang/protobuf/proto"
+	"io/ioutil"
+	"log"
+	"net/http"
+	urlutils "net/url"
+	"time"
+)
+
+type uploadResponse struct {
+	Error     []int `json:"500"`
+	Duplicate []int `json:"409"`
+	Success   []int `json:"201"`
+}
+
+//UploadBatch Called in CRON. Get keys from database and upload them to EFGS.
+func UploadBatch() {
+	ctx := context.Background()
+	logger := logging.FromContext(ctx)
+	currentTime := time.Now()
+	keys, err := database.GetDiagnosisKeys(currentTime.Format("2006-01-02"))
+	if err != nil {
+		logger.Errorf("Downloading keys error: %s", err)
+	}
+
+	if len(keys) <= 0 {
+		logger.Warn("No new keys in database.")
+		return
+	}
+
+	sortSentKeys(keys)
+	var diagnosisKeys []*DiagnosisKey
+	for _, k := range keys {
+		diagnosisKeys = append(diagnosisKeys, k.ToData())
+	}
+
+	sortDiagnosisKey(diagnosisKeys)
+	batch := makeBatch(diagnosisKeys)
+	batchTag := currentTime.Format("2006-01-02") + b64.StdEncoding.EncodeToString(batchToBytes(&batch))[:4]
+	logger.Debugf("Uploading batch (%d keys) with tag %s", len(keys), batchTag)
+	resp, err := uploadBatch(ctx, &batch, batchTag)
+	if err != nil {
+		if resp != nil {
+			for _, e := range resp.Error {
+				keys = append(keys[:e], keys[e+1:]...)
+			}
+		} else {
+			logger.Errorf("Batch uploading failed: %s", err)
+			return
+		}
+	}
+	if err := database.RemoveDiagnosisKey(keys); err != nil {
+		logger.Errorf("Removing uploaded keys from database failed: %s", err)
+		return
+	}
+}
+
+func uploadBatch(ctx context.Context, batch *DiagnosisKeyBatch, batchTag string) (*uploadResponse, error) {
+	logger := logging.FromContext(ctx)
+	secretsClient := secrets.Client{}
+
+	nbtlsPair, err := efgsutils.LoadX509KeyPair(ctx, efgsutils.NBTLS)
+	if err != nil {
+		logger.Errorf("Error loading authentication certificate: %v", err)
+		return nil, err
+	}
+
+	client, err := efgsutils.NewEFGSClient(ctx, nbtlsPair)
+	if err != nil {
+		logger.Errorf("Could not create EFGS client: %v", err)
+		return nil, err
+	}
+
+	var req *http.Request
+
+	efgsRootURL, err := secretsClient.Get("efgs-test-url")
+	if err != nil {
+		return nil, err
+	}
+
+	url, err := urlutils.Parse(string(efgsRootURL))
+	if err != nil {
+		return nil, err
+	}
+
+	url.Path = "diagnosiskeys/upload"
```

Comment:
I believe this configuration should be done before you start anything else because, without it, you can do nothing but fail.  
If you don't want to have a gazillion of params of this function, just wrap them all into a struct.